Ladybird/Tab.cpp

```diff
@@ -247,6 +264,49 @@ int Tab::tab_index()
     return m_window->tab_index(this);
 }
 
+void Tab::add_bookmark()
+{
+    auto model = qobject_cast<BookmarksModel*>(m_bookmarks_bar->model());
+    if (!model)
+        return;
+
+    model->add(m_location_edit->text());
+}
+
+void Tab::bookmark_clicked(QModelIndex const& index)
+{
+    auto model = qobject_cast<BookmarksModel*>(m_bookmarks_bar->model());
+    if (!model)
+        return;
+
+    auto url = model->data(index, Qt::DisplayRole).toString();
+    navigate(url);
+}
+
+void Tab::show_context_menu_bookmark(QPoint const& pos)
+{
+    auto list_view = qobject_cast<QListView*>(sender());
+    if (!list_view)
+        return;
+
+    auto bm = qobject_cast<BookmarksModel*>(list_view->model());
+    if (!bm)
+        return;
+
+    // Don't show context menu when there aren't bookmarks.
+    if (bm->get_urls().isEmpty())
+        return;
+
+    auto idx = list_view->indexAt(pos);
+
+    QMenu contextMenu("Context menu", this);
+    QAction remove_action("Remove", this);
+    QObject::connect(&remove_action, &QAction::triggered, this, [bm, idx] { bm->remove(idx.row()); });
+
+    contextMenu.addAction(&remove_action);
```

Comment:
It looks like the `remove_action` will go out of scope after this function? Is it safe to pass a pointer to it?

Reply:
It's safe because it is just the "moment" when context menu appears, if you click outside make sense all stuff are deallocated